### PR TITLE
Save oauth tokens to the openfaas config file

### PR DESCRIPTION
**Pull request by viveksyngh**
This commit saves oauth2 tokens to the openfaas config file with a
different auth type. The different auth type helps to determine whether
it is basic auth token or oauth2 token. It also refactors existing
config file function to cater the needs of auth command.

Fixes: #663

Signed-off-by: Vivek Singh <vivekkmr45@yahoo.in>

## Description

In this PR I have implemented the saving oauth2 tokens to the config file as per discussed architecture and also added the support to the `faas-cli list` command. It this approach looks, I can go ahead and implement the same in other commands.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

#663 

## How Has This Been Tested?

I have tested it on my local with a gateway which has oauth2 enabled.

```
./faas-cli auth --auth-url https://dev-cqxdj5cl.auth0.com/authorize --audience https://dev-cqxdj5cl.auth0.com/api/v2/ --client-id "4cCLcJDzkvQ2aQ11aHt7iLpbVfuPDOEk" -g http://gw.viveksyngh.me
Starting local token server on port 31111
Launching browser: https://dev-cqxdj5cl.auth0.com/authorize?audience=https%3A%2F%2Fdev-cqxdj5cl.auth0.com%2Fapi%2Fv2%2F&client_id=4cCLcJDzkvQ2aQ11aHt7iLpbVfuPDOEk&nonce=1571748418287862000&redirect_uri=http%3A%2F%2F127.0.0.1%3A31111%2Foauth%2Fcallback&response_type=token&scope=&state=1571748418287860000
credentials saved for http://gw.viveksyngh.me
Example:
	./faas-cli list --gateway "http://gw.viveksyngh.me" --token "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik9UVTFOMFV3TVRVMVFURk..
➜  faas-cli git:(add-ouath-token-to-config) ✗ cat ~/.openfaas/config.yml
auths:
- gateway: http://gw.viveksyngh.me
  auth: oauth2
  token: eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik9UVTFOMFV3TVR....
➜  faas-cli git:(add-ouath-token-to-config) ✗ ./faas-cli list -g http://gw.viveksyngh.me
Function                      	Invocations    	Replicas
cows                          	0              	1
➜  faas-cli git:(add-ouath-token-to-config) ✗ ./faas-cli login -u admin -p $PASSWORD -g http://gw.viveksyngh.me
WARNING! Using --password is insecure, consider using: cat ~/faas_pass.txt | faas-cli login -u user --password-stdin
Calling the OpenFaaS server to validate the credentials...
credentials saved for admin http://gw.viveksyngh.me
➜  faas-cli git:(add-ouath-token-to-config) ✗ cat ~/.openfaas/config.yml
auths:
- gateway: http://gw.viveksyngh.me
  auth: basic
  token: YWRtaW46ZWE4ZjM2NDFhMzYwMDliOGVmMTg4OWUzM2IyZDVmYWUzNjg2ZmY4Mg==
➜  faas-cli git:(add-ouath-token-to-config) ✗ ./faas-cli list -g http://gw.viveksyngh.me
Function                      	Invocations    	Replicas
cows                          	0              	1
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: proxy/auth.go [hunk 1]
```diff
@@ -11,16 +11,53 @@ import (
...
+func AddAuth(req *http.Request, gateway string) {
+	authConfig, err := config.LookupAuthConfig(gateway)
+	if err != nil {
+		// no auth info found
+		return
+	}
+
+	switch authConfig.Auth {
+	case config.BasicAuthType:
+		SetBasicAuth(req, authConfig)
```
Comment (LucasRoesler) [inline proxy/auth.go]: with this implementation, it seems the the `SetBasicAuth` and `SetOauth2` methods could be private.  If we are willing to change the signature, we could pass  the token

```
func AddAuth(req *http.Request, gateway string, token string) {}
```

and set the token auth here too, if it is non-empty.  But that would be reasonable to do in a future PR.

File: commands/auth.go [hunk 2]
```diff
@@ -187,6 +188,10 @@ func authClientCredentials() error {
 			return errors.Wrapf(tokenErr, "unable to unmarshal token: %s", string(tokenData))
 		}
 
+		if err := config.UpdateAuthConfig(gateway, token.AccessToken, config.Oauth2AuthType); err != nil {
```
Comment (alexellis) [inline commands/auth.go]: A nit, but this might be better as `OAuth`

File: commands/auth.go [hunk 3]
```diff
@@ -187,6 +188,10 @@ func authClientCredentials() error {
...
+			return err
+		}
+		fmt.Println("credentials saved for", gateway)
```
Comment (alexellis) [inline commands/auth.go]: Can we use Printf %s please?

File: commands/auth.go [hunk 4]
```diff
@@ -226,6 +231,11 @@ func makeCallbackHandler(cancel context.CancelFunc) func(w http.ResponseWriter,
 			}
 
 			if token := q.Get("access_token"); len(token) > 0 {
+
+				if err := config.UpdateAuthConfig(gateway, token, config.Oauth2AuthType); err != nil {
+					fmt.Printf("error while saving authentication token: %s", err.Error())
```
Comment (alexellis) [inline commands/auth.go]: Need \n at the end of the Printf

File: commands/auth.go [hunk 5]
```diff
@@ -226,6 +231,11 @@ func makeCallbackHandler(cancel context.CancelFunc) func(w http.ResponseWriter,
...
+				}
+				fmt.Println("credentials saved for", gateway)
```
Comment (alexellis) [inline commands/auth.go]: Let's use printf

File: proxy/list.go [hunk 6]
```diff
@@ -39,7 +39,7 @@ func ListFunctionsToken(gateway string, tlsInsecure bool, token string, namespac
 	if len(token) > 0 {
 		SetToken(getRequest, token)
 	} else {
-		SetAuth(getRequest, gateway)
+		AddAuth(getRequest, gateway)
```
Comment (alexellis) [inline proxy/list.go]: Why was this renamed?

File: commands/auth.go (hunk 3 above)
Comment (viveksyngh) [reply commands/auth.go]: Will update that.

File: proxy/list.go (hunk 6 above)
Comment (viveksyngh) [reply proxy/list.go]: It was not renamed but basically we are using a different function which is checks whether it is `oauth2` or `basic` auth and add the corresponding header. I have to update this method to all commands .. I just did this for `faas-cli list`  command.

File: commands/auth.go (hunk 5 above)
Comment (viveksyngh) [reply commands/auth.go]: Sure.

File: proxy/list.go (hunk 6 above)
Comment (viveksyngh) [reply proxy/list.go]: We can also update the `SetAuth` method and we will not have to add the support to each command separately. 

Since it is already being called inside each command but I rather chose to use a separate method `addAuth` and add the support gradually.
